master/internal/api_runs.go

```diff
@@ -43,3 +56,178 @@ func (a *apiServer) RunPrepareForReporting(
+		ColumnExpr("CASE WHEN e.parent_id IS NULL THEN NULL ELSE " +
+			"json_build_object('value', e.parent_id) END AS forked_from").
+		ColumnExpr("CASE WHEN e.progress IS NULL THEN NULL ELSE " +
+			"json_build_object('value', e.progress) END AS progress").
+		ColumnExpr("e.id AS experiment_id").
+		Column("e.owner_id").
+		ColumnExpr("e.config->>'description' AS description").
+		ColumnExpr("e.config->'resources'->>'resource_pool' AS resource_pool").
+		ColumnExpr("e.config->'searcher'->>'name' AS searcher_type").
+		ColumnExpr("e.config->'searcher'->>'metric' AS searcher_metric").
+		ColumnExpr("e.config->>'name' as experiment_name").
+		ColumnExpr("w.id AS workspace_id").
+		ColumnExpr("w.name AS workspace_name").
+		ColumnExpr("(w.archived OR p.archived) AS parent_archived").
+		Column("e.unmanaged").
+		ColumnExpr("p.name AS project_name").
+		Join("LEFT JOIN experiments AS e ON r.experiment_id=e.id").
+		Join("LEFT JOIN users u ON e.owner_id = u.id").
+		Join("LEFT JOIN projects p ON e.project_id = p.id").
+		Join("LEFT JOIN workspaces w ON p.workspace_id = w.id")
+}
+
+func sortRuns(sortString *string, runQuery *bun.SelectQuery) error {
+	if sortString == nil {
+		return nil
+	}
+	sortByMap := map[string]string{
+		"asc":  "ASC",
+		"desc": "DESC NULLS LAST",
+	}
+	orderColMap := map[string]string{
+		"id":                   "id",
+		"description":          "description",
+		"name":                 "name",
+		"searcherType":         "searcher_type",
+		"searcherMetric":       "searcher_metric",
+		"startTime":            "r.start_time",
+		"endTime":              "r.end_time",
+		"state":                "r.state",
+		"progress":             "COALESCE(progress, 0)",
+		"user":                 "display_name",
+		"forkedFrom":           "e.parent_id",
+		"resourcePool":         "resource_pool",
+		"projectId":            "e.project_id",
+		"checkpointSize":       "checkpoint_size",
+		"checkpointCount":      "checkpoint_count",
+		"duration":             "duration",
+		"searcherMetricsVal":   "r.searcher_metric_val",
+		"externalExperimentId": "e.external_experiment_id",
+		"externalRunId":        "r.external_run_id",
+		"experimentId":         "e.id",
+		"numRuns":              "num_runs",
+	}
+	sortParams := strings.Split(*sortString, ",")
+	hasIDSort := false
+	for _, sortParam := range sortParams {
+		paramDetail := strings.Split(sortParam, "=")
+		if len(paramDetail) != 2 {
+			return status.Errorf(codes.InvalidArgument, "invalid sort parameter: %s", sortParam)
+		}
+		if _, ok := sortByMap[paramDetail[1]]; !ok {
+			return status.Errorf(codes.InvalidArgument, "invalid sort direction: %s", paramDetail[1])
+		}
+		sortDirection := sortByMap[paramDetail[1]]
+		switch {
+		case strings.HasPrefix(paramDetail[0], "hp."):
+			param := strings.ReplaceAll(paramDetail[0], "'", "")
+			hps := strings.ReplaceAll(strings.TrimPrefix(param, "hp."), ".", "'->'")
+			runQuery.OrderExpr(
+				fmt.Sprintf("r.hparams->'%s' %s", hps, sortDirection))
```

Comment:
this feels very close to being vulnerable to sql injection, can we write this this without using ```fmt.Sprintf```?